### PR TITLE
Initial draft of IANA port assignment form

### PR DESCRIPTION
**Pull request by BlueHatbRit**
This is the initial draft of the IANA port assignment form. As you might be able to tell, I've not filled one of these out before so I may have missed the point on some of these questions, hence the request for comments and feedback.

In the markdown document I've only listed the questions that we are required to answer. Please see the full form here - https://www.iana.org/form/ports-services to see any questions that have been intentionally skipped. Would be great if someone could also check this in case I've missed something important!

Many of the questions refer to the protocol as opposed to the service.

ccing for review / input: @mheffner, @mrtazz, @lzap. Please feel free to pull in anyone else who might have experience and be interested in helping out.

### CODE REVIEW

File: docs/iana-application.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+The primary statsd server implementation maintains support for all current and LTS versions of nodejs. The protocol itself is small enough to enable the addition of authentication tokens should it fit the use case in the future via purely additive means.
+
+### Please explain why a unique port assignment is necessary as opposed to a port in range (49152-65535) or existing port.
+Statsd is an industry standard application used by many to aid in the development and monitoring of running applications. Statsd's popularity has lead to many guides and users making use of the default port 8125. Assignment of a unique port will enable operating system providers to reserve the port for statsd usage to help safe guard against other services conflicting with the statsd configuration and attempting to bind to the port, and potentially interrupting critical metric collection.
+
+### Please explain the state of development of your protocol.
+The statsd network protocol has been well defined for several years and used by many application developers. The protocl is only expected to be added to in the future, not broken or changed in a way that would not support backwards compatibility.
```
Comment (lzap) [inline docs/iana-application.md]: Probably "a decade" now? Isn't the original from like 2005 or something? ;)

File: docs/iana-application.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+Please provide any other information that would be helpful in understanding how this protocol differs from existing assigned services.
+Statsd is an open source project, it's documentation implementation can be found at https://github.com/statsd/statsd.
```
Comment (lzap) [inline docs/iana-application.md]: I'd also include few major implementations, so it's pretty obvious this is not some one-man-show. :-)

File: docs/iana-application.md (hunk 1 above)
Comment (BlueHatbRit) [reply docs/iana-application.md]: That's a very fair point, I can also see a typo on this line which I'll fix.

File: docs/iana-application.md (hunk 2 above)
Comment (BlueHatbRit) [reply docs/iana-application.md]: Great idea, I'll toss in a few library links and some of the other server implementations as well.

File: docs/iana-application.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+### Description
+Network daemon for collecting metrics
+
+### Reference
+StatsD is used to collect statistics such as counters and timers to measure system and application performance. StatsD receives metrics over UDP or TCP in a basic line protocl of the following format:
+
+`<metricname>:<value>|<type>`
+
+The metric name (or "bucket") is an identifier for the metric being sent. Values represent the metric value being pushed to statsd and should be of the specified "type". Statsd supports numerous mertric types such as counters, gagues, and timers. Each has it's own single character key which is used in the type section.
```
Comment (mheffner) [inline docs/iana-application.md]: Should this mention expansions that include Dogstatsd formats?

File: docs/iana-application.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+Metrics are collected and flushed at a desired interval to the plugable backend for storage and analysis.
+
+### If UDP is requested, please explain how traffic is limited, and whether the protocol reacts to congestion.
+Statsd does not react to congestion itself, but does provide the means for creating statsd clusters to aid performance.
+
+### If UDP is requested, please indicate whether the service is solely for the discovery of hosts supporting this protocol.
+Statsd is not intended for service discovery.
+
+### Please explain how your protocol supports versioning.
+As new metric types are introduced they will be assigned their own "short code". Statsd is intended to be run internally and not exposed to the wider world and so ensuring the statsd version supports the desired metric type is the responsibility of the administrator.
```
Comment (mheffner) [inline docs/iana-application.md]: The earlier format used "type" as the metric type, whereas this uses "short code" as the type. Should we use a consistent naming for the type?

File: docs/iana-application.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+### If your request is for more than one transport, please explain in detail how the protocol differs over each transport.
+Statsd defaults to UDP but can be configured to use TCP to ensure delivery of packets. The protocol's contracts across TCP is the same as across UDP.
+
+### Please describe how your protocol supports security. Note that new services are expected to support security capabilities and to avoid insecure variants.
+Statsd can be run behind authentication services which can be errected to proxy traffic through to statsd if users wish to expose it across public networks such as the internet.
+
+The primary statsd server implementation maintains support for all current and LTS versions of nodejs. The protocol itself is small enough to enable the addition of authentication tokens should it fit the use case in the future via purely additive means.
```
Comment (mheffner) [inline docs/iana-application.md]: Should we add that a recommended mode of operation is to run statsd on localhost and rarely, if ever, expose to the wider internet?

File: docs/iana-application.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+### Please explain the state of development of your protocol.
+The statsd network protocol has been well defined for several years and used by many application developers. The protocl is only expected to be added to in the future, not broken or changed in a way that would not support backwards compatibility.
+
+Please provide any other information that would be helpful in understanding how this protocol differs from existing assigned services.
+Statsd is an open source project, it's documentation implementation can be found at https://github.com/statsd/statsd.
+
+The project has existed for around 9 years and has emerged as popular part of server side web application developer toolchains. Many other open source services make use of statsd and assume it's presence in an application monitoring stack, such as Graphite and Prometheus.
```
Comment (mheffner) [inline docs/iana-application.md]: typo: `it's` -> `its`

File: docs/iana-application.md [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+- [ ] DCCP
+
+### Service Name
+StatsD
+
+### Desired Port Number
+8125
+
+### Description
+Network daemon for collecting metrics
```
Comment (mheffner) [inline docs/iana-application.md]: "...and publishing to various metric collection systems or services"?

File: docs/iana-application.md [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+### Reference
+StatsD is used to collect statistics such as counters and timers to measure system and application performance. StatsD receives metrics over UDP or TCP in a basic line protocl of the following format:
```
Comment (mheffner) [inline docs/iana-application.md]: typo: `protocl`

File: docs/iana-application.md (hunk 3 above)
Comment (BlueHatbRit) [reply docs/iana-application.md]: Good idea, I forgot about dogstatsd, that'll help outline the stability of the protocol as well.

File: docs/iana-application.md (hunk 4 above)
Comment (BlueHatbRit) [reply docs/iana-application.md]: Yes you're quite right, I'll make that change

File: docs/iana-application.md (hunk 5 above)
Comment (BlueHatbRit) [reply docs/iana-application.md]: Yes I think that's a good idea, I'll whip up a few lines about that as well

**Comment by BlueHatbRit** [comment]
I've made the suggested changes, if anyone is able to re-review and make sure they're happy then that'd be awesome 😄

File: docs/iana-application.md [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+### Description
+Network daemon for collecting metrics and publishing to various metric aggregation systems or services.
+
+### Reference
+StatsD is used to collect statistics such as counters and timers to measure system and application performance. StatsD receives metrics over UDP or TCP in a basic line protocl of the following format:
+
+`<metricname>:<value>|<type>`
+
+The metric name (or "bucket") is an identifier for the metric being sent. Values represent the metric value being pushed to StatsD and should be of the specified "type". Statsd supports numerous mertric types such as counters, gagues, and timers. Each has it's own single character "short code" key which is used in the type section.
```
Comment (mheffner) [inline docs/iana-application.md]: `it's` -> `its`

File: docs/iana-application.md [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+Metrics are collected and flushed at a desired interval to the plugable backend for storage and analysis.
+
+Other projects and organisations have also created extensions to the StatsD protocol, one popular example being DogStatsD by DataDog. See https://docs.datadoghq.com/developers/dogstatsd/ for more details on the extension.
+
+### If UDP is requested, please explain how traffic is limited, and whether the protocol reacts to congestion.
+Statsd does not react to congestion itself, but does provide the means for creating StatsD clusters to aid performance.
```
Comment (mheffner) [inline docs/iana-application.md]: Consistently upcase the `D` in `StatsD`?

Comment (BlueHatbRit) [reply docs/iana-application.md]: My fine and replace foo failed me

**Comment by BlueHatbRit** [comment]
I'll be honest I have no idea if we can resubmit after, but I'd assume we can so we'll have to ride the boat and see!

**Comment by BlueHatbRit** [comment]
The application has been submitted for StatsD using the `maintainers@statsd.app` email address. Hopefully we'll hear back soon 🚀

**Comment by BlueHatbRit** [comment]
Application id is `1152573` for anyone that wants to keep a watch on it.

**Comment by BlueHatbRit** [comment]
Sadly the application was declined, they had some specific issues with the application itself but also with the general use case. By the sounds of the response even a resubmission wouldn't really fit the port assignment use case for them. I'll post the specifics when I get back to my computer in a few days but the outlook doesn't look very promising.

**Comment by BlueHatbRit** [comment]
Below is the body of the email received by the IANA Services Specialist. It explains the reasons the IESG Expert gave for rejecting the application.

> Dear Elliot, 
>
> The expert has reviewed your application for a port number. He had the following comments: 
>
> Because this service is intended to operate internal to a machine, it would not be eligible for an assignment.
>
> This request would also have been declined for any of the following criteria, FYI:
>
> The concern about seeking an assigned port to avoid squatting (unauthorized use of port 8125 by those not assigned it) would not be alleviated by the assignment of a new port. In fact, this application is an example of such squatting in its current use of that port anyway, so it is curious why this application should assume protection against the behavior it itself already exhibits.
>
> Services over UDP are expected to follow the guidelines of RFC8085. This is not consistent with the response to question 2 below.
>
>New services are expected to support security to avoid the need for additional ports in the future for new versions/variants. This is not consistent with the response to question 4 below; it is not sufficient to put the onus on the user to determine whether things run using locally available information not inside the service.
>
> New services are expected to provide security. It is not sufficient to put this responsibility on the user, as per the response to question 6 below.
>
> Finally, if this were to proceed, at best the assignment would only proceed for the currently used port (8125), given it is unassigned. This would not alleviate your concerns and the rationale for this request and so a new assignment would not be productive.

It sounds to me like they're pretty set against assigning a port to statsd's use case, unless anyone interprets this differently and thinks it's worth persuing more, I'll probably look to get the request closed by IANA.

**Comment by lzap** [comment]
At least you've tried, /cc @natoscott I believe SELinux team will have no problems with assigning port type to statsd. Thanks for info!

**Comment by BlueHatbRit** [comment]
Thanks @lzap, please do let us know if this does happen, we'd be keen to know 😄

**Comment by BlueHatbRit** [comment]
Closing as this was rejected by IANA/IESG. Would be great to see if SELinux can assign a port anyway, I'll leave the origina issue #682 open for now.

**Comment by lzap** [comment]
@BlueHatbRit well it actually happened already in Fedora:

https://github.com/fedora-selinux/selinux-policy/commit/5fa8f7c0810822ecb5fc51b8d02da00eab6c730a
https://github.com/fedora-selinux/selinux-policy-contrib/commit/5e590a3b47ad3aa3bb6566cd5c188bd052658d7e

And I asked my colleagues to backport this into RHEL8 line.

https://bugzilla.redhat.com/show_bug.cgi?id=1746511

Thanks!

**Comment by saper** [comment]
I had a quick look at this and I think it could be salvaged (glad to help) but one thing that really worries me is complete lack of security. Versioning is also problematic.

There is also an interesting problem that this is a one-way protocol and here we specify only the server-side behaviour. In this case it is pretty easy to flood the destination with too much traffic. How do large installations using statsd cope with this, preventing burst traffic hitting a single instance?

**Comment by lzap** [comment]
> In this case it is pretty easy to flood the destination with too much traffic.

Have you tried yourself? In case of UDP there is a buffer and all excessive traffic is simply dropped by the kernel. In case of TCP, the traffic will be slowed down by the protocol as it's guaranteed delivery. I believe statsd protocol/service is not different from pretty much any other service.

But I guess it's the wrong thread for this discussion.

**Comment by BlueHatbRit** [comment]
I don't have the exact statistics, but I'd wager than 90% of installations use UDP, TCP feels very much like an addition just "because we can". As @lzap said with UDP the traffic is just dropped.

Since StatsD is designed to be run internally and never be exposed to the internet, a lot of the responsibility is on the user to understand the load the spin up a new node as required if they wish to handle the load. If the sys admin wishes to scale up to manage additional load then there are systems out there for autoscaling but that really sits on an infrastructure level rather than with statsd itself.

**Pull request closed without merging**